src/lib/bootstrap_controller/bootstrap_controller.ml

```diff
@@ -90,62 +99,77 @@ module Make (Inputs : Inputs_intf) :
   let done_syncing_root t =
     Option.is_some (Syncable_ledger.peek_valid_tree t.syncable_ledger)
 
-  let length protocol_state =
-    Consensus.Mechanism.Protocol_state.consensus_state protocol_state
+  let length external_transition =
+    external_transition |> External_transition.Verified.protocol_state
+    |> Consensus.Mechanism.Protocol_state.consensus_state
     |> Consensus.Mechanism.Consensus_state.length |> Coda_numbers.Length.to_int
 
-  let on_transition t (transition, time_received) =
+  let on_transition t (candidate_transition : External_transition.Verified.t) =
     let module Protocol_state = Consensus.Mechanism.Protocol_state in
-    let candidate = External_transition.protocol_state transition in
-    let previous_state_hash = Protocol_state.previous_state_hash candidate in
+    let candidate_state =
+      External_transition.Verified.protocol_state candidate_transition
+    in
+    let previous_state_hash =
+      Protocol_state.previous_state_hash candidate_state
+    in
     let input : Ancestor.Input.t =
       { descendant= previous_state_hash
-      ; generations= length candidate - length t.best_with_root.root }
+      ; generations= length candidate_transition - length t.current_root }
     in
-    if
-      not
-        (Consensus.Mechanism.received_at_valid_time
-           (Protocol_state.consensus_state candidate)
-           ~time_received)
-    then (
-      Logger.faulty_peer t.logger
-        "received protocol state at invalid time while bootstrapping" ;
-      Deferred.unit )
-    else if done_syncing_root t || (not @@ worth_getting_root t candidate) then
-      Deferred.unit
+    if done_syncing_root t || (not @@ worth_getting_root t candidate_state)
+    then Deferred.unit
     else
-      match%map
+      match%bind
         Network.get_ancestry t.network (input.descendant, input.generations)
       with
       | Error e ->
-          Logger.error t.logger
-            !"Could not get the proof of ancestors from the \
-              network:%{sexp:Error.t}"
-            e
-      | Ok (ancestor, proof) -> (
-          let ancestor_length =
-            Protocol_state.(Consensus_state.length (consensus_state ancestor))
+          Deferred.return
+          @@ Logger.error t.logger
+               !"Could not get the proof of ancestors from the \
+                 network:%{sexp:Error.t}"
+               e
+      | Ok (ancestor_transition, proof) -> (
+          let result =
+            let open Deferred.Or_error.Let_syntax in
+            let%bind verified_ancestor_transition =
+              Protocol_state_validator.validate_proof ancestor_transition
```

Comment:
why is it safe not to validate consensus state here? Do we not have enough information at this point?

Reply:
how annoying would it be to have `External_transition.Proof_verified.t` which would mean "consensus state not verified, proof is verified" and then `External_transition.Verified.t` would mean "proof is verified and consensus state is verified"